support umask for git repos

bugfood opened the pull request:
When cloning a repo, it can be useful to have shared access such that
other members of the cloning user's group are able to manage the repo as
well. This requires two components:

1. The setgid bit to be set on directories; this can be accomplished by
   puppet code which creates the repo directory before calling vcsrepo.
2. A umask that allows group writeability to be used during the clone.

This patch implements the second component. Note that for git,
specifically, there is a core.sharedRepository option, but this only
seems to be respected on 'git init', not 'git clone'. By default, git
will respect the umask anyhow.

I do not have ready access to the other version control systems, so I
only implemented this for git, but using a umask is likely portable to
other version control systems as well.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/puppetlabs/puppetlabs-vcsrepo?pullRequest=541) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/puppetlabs/puppetlabs-vcsrepo?pullRequest=541) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/puppetlabs/puppetlabs-vcsrepo?pullRequest=541) it.</sub>

bugfood commented:
I can try to correct the test failures. I'm unsure about this one from rubocop:
```
lib/puppet/type/vcsrepo.rb:312:19: C: [Correctable] Style/RegexpLiteral: Use %r around regular expression. ([https://rubystyle.guide#percent-r](https://rubystyle.guide/#percent-r))
      if value =~ /^0?[0-7]{1,4}$/
                  ^^^^^^^^^^^^^^^^
````

The linked-to style guide says:
> Use %r only for regular expressions matching at least one / character.

...so it seems like rubocop is telling me to violate the style guide.

chelnak commented:
Hey! Thank you for the updates.

I'm currently working on the acceptance tests in another stream as they seem a little unwell right now.. however i'll take a look at the failures here to make sure everything is working as expected.

bugfood commented:
> Hey! Thank you for the updates.
> 
> I'm currently working on the acceptance tests in another stream as they seem a little unwell right now.. however i'll take a look at the failures here to make sure everything is working as expected.

Nice, thank you. I can rebase this again later if you let me know when that is ready.

chelnak commented:
Hello! Any chance you could rebase again please?

bugfood commented:
> Hello! Any chance you could rebase again please?

Sure; done. There minor merge conflicts, but no recent changes should affect the code in the PR.

petergmurphy commented:
Hi @bugfood,

Would it be possible for you to rebase this PR again? Apologies for the inconvenience.

bugfood commented:
> Hi @bugfood,
> 
> Would it be possible for you to rebase this PR again? Apologies for the inconvenience.

@petergmurphy done.

There were merge conflicts caused by the introduction of the `skip_hooks` feature, but I don't see any recent code what would otherwise interfere.

bugfood commented:
The latest update should fix the test failures.

jordanbreen28 commented:
@bugfood unfortunately we still have quite a lot of test failures - when you have some time, can you look into this?

jordanbreen28 commented:
Hi @bugfood 
Our team is auditing are ageing PRs, and as this request has had no activity since August 18th, I feel that it may be best to close this, for now.
If in the future you have some time to pick it up again and are able address the comments above, please re-open this request or create a new one. :-)
Thanks for your understanding.

bugfood commented:
@jordanbreen28 I had some time to get to this today.

I reproduced the test failure locally--it turns out that my test was clashing with another test using the same directory for the checkout, and such directories aren't removed between tests.

I fixed the failure locally and updated my branch. Hopefully the CI test suite will pass now.

Can you please re-open this PR? I don't see how to do that--my account probably lacks the permissions.

jordanbreen28 commented:
Hi @bugfood - thank you for taking the time to investigate the test failures! Unfortunately I cannot re-open this request, as Github is flagging "The for-upstream branch was force-pushed or recreated". This seems to be a known issue with force-pushed PRs on GH (there's a thread about it [here](https://github.com/isaacs/github/issues/361)).

Can I ask that you try the steps detailed in the comment [here](https://github.com/isaacs/github/issues/361#issuecomment-561158372), or if that fails, please create a new PR and reference this one.
Apologies for any inconvenience!

bugfood commented:
Hmm, I didn't know about that bug. That workaround didn't work for me, nor did the steps here: https://github.com/isaacs/github/issues/361#issuecomment-219088644.

I think it might be possible my user still lacks the permissions to re-open PRs, but I'll open a new PR shortly, anyway, on a different branch. We can use that PR or (if possible) re-open this one and update this branch.
